[Gutenberg] Set consistent editor colours in both the WordPress and Jetpack apps

SiobhyB opened the pull request:
Fixes the iOS side of https://github.com/wordpress-mobile/gutenberg-mobile/issues/5047, [Android PR is here](https://github.com/wordpress-mobile/WordPress-Android/pull/16968/)

## Description

In the Jetpack app, some UI elements like the cursor, text selection, and caret are using a green colour. However, the rest of the editor’s elements are using blue. 

To be consistent with colours across the editor, this PR seeks to update them for consistency across both the WordPress and Jetpack apps. The PR specifically achieves the following for the Jetpack app:

* Updates the colour of UI elements within the main editor (e.g. cursor, highlight) from green to blue.
* Updates the accent of sub screens within Post Settings from green to blue.

There's still some known inconsistencies that need to be addressed in separate PRs, however:

* The media picker (available from media blocks and the Post Settings screen) still has some green.
* There is still green visible in other screens 'connected' to the editor, such as Preview, History, Help & Support, etc.

These remaining inconsistencies have been captured in the issue at https://github.com/wordpress-mobile/gutenberg-mobile/issues/4889 for the future.

## Screenshots

A before/after screenshot taken from the Jetpack app to show the editor's green → blue colour changes:

| Before  | After |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/180868507-303c44fb-23b4-474b-93fa-2c18d299f317.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/180877183-cec91c87-798f-42ab-b376-b1bd53e20640.png" width="100%"> |

A before/after taken from the WordPress app (dark mode) to confirm there are no unexpected changes:

| Before  | After |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/181541724-efb97952-004d-4974-b774-a6d7968b5710.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/181542827-627e6555-0d0f-4109-8876-c1fe1d9cbadf.png" width="100%"> |

## Testing

* Download this PR's installable build for both the WordPress and Jetpack apps.
* In each app, navigate to the editor.
* Verify the colours used throughout the editor appear consistent for both app. The cursor, text selection, and caret should all now be blue for Jetpack (this is in contrast to the green that's used outside of the app's editor). There should be no change in the WordPress app, with the colours remaining blue.  

Special considerations to keep in mind while testing include:

* This PR aims to bring colour consistency to the Post Settings screen as well as the editor.
* Testing of any popovers, modals, or less visible UI elements within the editor would be beneficial. 
* As this is a UI change, dark/light mode should be a consideration.
* These changes should be noticeable in both the post _and_ the page editor.

## Code Changes

The colours we're aiming to change are using the default `.tintColor`, which is set to `.primary` via the following line of code:

https://github.com/wordpress-mobile/WordPress-iOS/blob/3eabde78c08cd210dfe1ed26109279d32d448ad7/WordPress/Classes/Extensions/Colors%20and%20Styles/WPStyleGuide%2BApplicationStyles.swift#L20-L22

`.primary` is set as `.jetpackGreen` [here for the Jetpack app](https://github.com/wordpress-mobile/WordPress-iOS/blob/a96da1dbcf200eb532e83666fb12fca5c1812129/WordPress/Jetpack/AppStyleGuide.swift#L19) and `.blue` [here for the WordPress app](https://github.com/wordpress-mobile/WordPress-iOS/blob/a96da1dbcf200eb532e83666fb12fca5c1812129/WordPress/Classes/Utility/App%20Configuration/AppStyleGuide.swift#L18).

As a number of UI elements needed to be updated as part of this PR, a new `.editorPrimary` variable was created, following the same patterns found for other colours in [the `UIColor+MurielColors.swift` file](https://github.com/wordpress-mobile/WordPress-iOS/blob/d6b813b903c34d4d66ac07c5e452b378fd952e63/WordPress/Classes/Extensions/Colors%20and%20Styles/UIColor+MurielColors.swift), such as `.primary` and `.success`.

### Root React Native View

The `tintColor` for the editor's main view (i.e. the root React Native view) has been set here: 

https://github.com/wordpress-mobile/WordPress-iOS/blob/43b9994672179dc1c0203f46cfce2a36e333bbb5/WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift#L568

### React Native Modals

`configureDefaultTint()` is [setting the tint colour for the entire window](https://github.com/wordpress-mobile/WordPress-iOS/blob/3eabde78c08cd210dfe1ed26109279d32d448ad7/WordPress/Classes/Extensions/Colors%20and%20Styles/WPStyleGuide%2BApplicationStyles.swift#L20-L22) of each screen via [`UIWindow`](https://developer.apple.com/documentation/uikit/uiwindow). As React Native's modals exist outside of the root view, it's necessary to override the default `tintColor` once again to ensure consistency in `BottomSheets` and other parts of the app that may be using modals.

This was achieved on the native side via `override func present`:

https://github.com/wordpress-mobile/WordPress-iOS/blob/43b9994672179dc1c0203f46cfce2a36e333bbb5/WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift#L415-L420

### React Native Action Sheets and Switches

The default `.tintColor` for switches/toggles within the editor's modals needed to be overridden, which was done via the following line of code:

https://github.com/wordpress-mobile/WordPress-iOS/blob/43b9994672179dc1c0203f46cfce2a36e333bbb5/WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift#L576

Note, the above was set within `GutenbergViewController.swift` rather than `WordPressAppDelegate.swift` as `RCTModalHostViewController` resulted in an out-of-scope error in the latter's file.

In addition, the `.tintColor` for action sheets has been set directly in React Native via https://github.com/wordpress-mobile/gutenberg-mobile/pull/5074.

### Post Settings Screen

It was possible to set the `tintColor` for multiple UI elements in the Post Settings screen.

Some hacky code has been used for the parent category screen (found within Post Settings). This was necessary because the screen is used elsewhere in the app (via Site Settings → Category) and there wasn't any other views in the hierarchy that could be used to only target the Post Settings UI:

https://github.com/wordpress-mobile/WordPress-iOS/blob/56dab966a361844f1e30cdea8ce4d71218d0ae1b/WordPress/Classes/System/WordPressAppDelegate.swift#L889-L893


## Regression Notes
1. Potential unintended areas of impact

There's the potential for this PR to unexpectedly change colours in either the WordPress or Jetpack app beyond what's desired. For example, the cursor colour may unexpectedly change within the WordPress app. 

2. What I did to test those areas of impact (or what existing automated tests I relied on)

I manually tested both apps, making sure to experiment with various blocks and writing flows.

3. What automated tests I added (or what prevented me from doing so)

No automated tests.

<hr />

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

SiobhyB committed:
Update editor's tint colour to always be WP Blue

The editor's "tintColor" is explicitly set to be "wordPressBlue" with this commit. The most notable colour that will change following this is the editor's cursor and "highlight" colour when text is selected.

SiobhyB committed:
Update tintColor to match correct shade of blue

SiobhyB committed:
Merge branch 'trunk' into gutenberg/incorporate-consistent-tint-color

SiobhyB committed:
Update RELEASE-NOTES

dcalhoun commented on WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift:
```diff
@@ -554,6 +554,7 @@ class GutenbergViewController: UIViewController, PostEditor, FeaturedImageDelega
 extension GutenbergViewController {
     private func setupGutenbergView() {
         view.backgroundColor = .white
+        view.tintColor = UIColor.muriel(color: MurielColor(name: .blue))
```
In an effort to address the _Editor UI_ references I called out, I researched React Native's `selectionColor` property for `TextInput` a bit. Unfortunately, I was unable to identify where the [default value](https://github.com/facebook/react-native/search?q=selectionColor) for [`selectionColor`](https://reactnative.dev/docs/textinput#selectioncolor) is sourced or why this `tintColor` does not influence `TextInput` or [`Switch`](https://reactnative.dev/docs/switch#ios_backgroundcolor-ios).

dcalhoun reviewed (commented):
Thank you for investigating this. This change makes sense to me, however, I noticed several UI instances where this change does not control. 

The _Editor UI_ references below are bit unexpected IMO. If it is easily addressable, then I think it makes sense to avoid the Jetpack brand green for these places in service of wordpress-mobile/gutenberg-mobile#5047. 

<details><summary>Editor UI</summary>

| Location | Image |
| - | - |
| Inserter Search | <img alt="Inserter search displays unexepected Jetpack brand green" src="https://user-images.githubusercontent.com/438664/181034737-b5652500-9445-4107-b62e-716a11c0477a.jpeg" width="100" /> |
| Link Picker | <img alt="Link picker displays unexepected Jetpack brand green" src="https://user-images.githubusercontent.com/438664/181034852-f560c1b0-290f-40df-9652-44f03e10e6ea.jpeg" width="100" /> |
| Text Area | <img alt="Bottom sheet text area displays unexepected Jetpack brand green" src="https://user-images.githubusercontent.com/438664/181034857-5170e975-54f6-4bfc-b542-3dbdfe8747b0.jpeg" width="100" /> |
| Text Input | <img alt="Bottom sheet text input displays unexepected Jetpack brand green" src="https://user-images.githubusercontent.com/438664/181034861-324718a1-2d03-4d72-8303-07dbec15c50f.jpeg" width="100" /> |
| Range Input | <img alt="Bottom sheet range input displays unexepected Jetpack brand green" src="https://user-images.githubusercontent.com/438664/181034858-5a995426-55bd-4061-a330-5d8d9bf695fc.jpeg" width="100" /> |
| Toggle | <img alt="Bottom sheet toggle displays unexepected Jetpack brand green" src="https://user-images.githubusercontent.com/438664/181034865-8b77decb-e4b3-40b5-86f0-41a4f3ec1bc5.jpeg" width="100" /> |

</details>

The _Post Settings UI_ references below are bit of a "gray area" IMO, I think there is an argument to make that this UI is not within the editor and should mirror the general Jetpack app color scheme. We can likely leave the _Post Settings UI_ as-is.

<details><summary>Post Settings UI</summary>

| Location | Image |
| - | - |
| Post Slug | <img alt="Post settings slug input displays unexepected Jetpack brand green" src="https://user-images.githubusercontent.com/438664/181034868-646641e6-59b3-48b4-8cf4-8f54175798bc.jpeg" width="100" /> |
| Post Excerpt | <img alt="Post settings excerpt input displays unexepected Jetpack brand green" src="https://user-images.githubusercontent.com/438664/181034870-e370296a-164a-4100-89ce-6b935e6f5c91.jpeg" width="100" /> |

</details>

I could be convinced this PR is worth merging as-is to address the blatant issue of rich text fields utilizing the unexpected Jetpack brand green, with a follow up for the _Editor UI_ references. WDYT about these references?

SiobhyB committed:
Set tintColor for *entire* editor window

SiobhyB committed:
Clarify purpose of code in comment

SiobhyB committed:
Tidy up code for clarity and conciseness

SiobhyB committed:
Merge branch 'trunk' into gutenberg/incorporate-consistent-tint-color

SiobhyB commented:
Thanks so much for spotting those UI pieces that the original PR missed, @dcalhoun!

I found that this is caused by `configureDefaultTint()` [setting the tint colour for the entire window](https://github.com/wordpress-mobile/WordPress-iOS/blob/3eabde78c08cd210dfe1ed26109279d32d448ad7/WordPress/Classes/Extensions/Colors%20and%20Styles/WPStyleGuide%2BApplicationStyles.swift#L20-L22) of each screen via [`UIWindow`](https://developer.apple.com/documentation/uikit/uiwindow). So, setting the tint colour specifically for the editor view isn't enough, it needed to be set for the window. I went ahead to implement that in https://github.com/wordpress-mobile/WordPress-iOS/pull/19113/commits/38aac84dfa9d20d79012a0a6bcc23a56f1a587d7. 

In my testing, this addresses all the views from your screenshots, including the Post Editor settings. However, I see it also changes the tint colour for other areas of the app after exiting the editor. 🤔 

I'll work on this further tomorrow and update you when it's ready for review again. 🙇‍♀️

dcalhoun commented:
> I found that this is caused by `configureDefaultTint()` [setting the tint colour for the entire window](https://github.com/wordpress-mobile/WordPress-iOS/blob/3eabde78c08cd210dfe1ed26109279d32d448ad7/WordPress/Classes/Extensions/Colors%20and%20Styles/WPStyleGuide%2BApplicationStyles.swift#L20-L22) of each screen via [`UIWindow`](https://developer.apple.com/documentation/uikit/uiwindow).

Given that all the references I called out are rendered within a modal, I wonder if that explains why modifying the Gutenberg root view does not impact them. My understanding is that React Native `Modal` is rendered _outside of the default root view_. I wonder if there is a way to modify the `tintColor` for the React Native `Modal`. 🤔 

> However, I see it also changes the tint colour for other areas of the app after exiting the editor. 🤔

I do not find this surprising. The current code mutates the top-level window `tintColor` when Gutenberg's view is created, but never resets it. We would likely need to reset the `tintView` when Gutenberg is dismissed. However, this approach concerns me a bit. I would want to seek input from a more experience iOS developer on this approach.

SiobhyB committed:
Revert previous changes

Setting the tintColor to the entire window caused colours to be set elsewhere in the app. As such, this commit, reverts that change and limits the tintColor to the editor's root view.

SiobhyB committed:
Set tintColor for modals

SiobhyB committed:
Merge branch 'trunk' into gutenberg/incorporate-consistent-tint-color

SiobhyB committed:
Set tintColor to Post Settings screen

SiobhyB committed:
Set tintColor for last Post Setting UI elements

SiobhyB committed:
Add editorPrimary var for clarity and reusability

SiobhyB committed:
Address mistake in previous commit

In the previous commit, I accidentally committed the reference to "editorPrimaryLight", but hadn't intended to. This commit removes that reference and adds back the "editorPrimary" function that was accidentally deleted.

SiobhyB committed:
Set tintColor on Post Settings' date picker

SiobhyB committed:
Small tweaks to code positioning, for readability

SiobhyB commented on WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift:
```diff
@@ -408,6 +408,11 @@ class GutenbergViewController: UIViewController, PostEditor, FeaturedImageDelega
         }
     }
 
+    override func present(_ viewControllerToPresent: UIViewController, animated flag: Bool, completion: (() -> Void)? = nil) {
```
[react-native-modal](https://github.com/react-native-modal/react-native-modal) (the library Gutenberg's using to display modals) leverages iOS' modal presentation. The method described in [this article](https://medium.com/bumble-tech/intercepting-and-redefining-modal-presentation-from-a-third-party-framework-e91079342443) was the only way I was able to override this, though I'm not sure if this would be considered too hacky.

dcalhoun replied:
I do not consider this approach to be a hack. Based on my understanding, I think it is valid to hook into `present` to modify the presentation of the modals. 👍🏻

SiobhyB commented on WordPress/Classes/ViewRelated/Post/Scheduling/SchedulingDatePickerViewController.swift:
```diff
@@ -211,5 +211,7 @@ extension DatePickerSheet where Self: UIViewController {
 
         view.pinSubviewToSafeArea(stackView)
+
+        view.tintColor = UIColor.editorPrimary
```
Set the `tintColor` for the calendar available from Post Settings. I wasn't able to find a clean way to change this directly from `PostSettingsViewController.m` and don't believe the calendar is used outside of the editor.

dcalhoun replied:
I suppose this is relatively safe given the view controller resides within `Post` and is specifically named `SchedulingDate`. As you alluded, this change does inhibit reusing this view controller outside of the editor. I am unsure if that is a big concern currently. It is likely alright to utilize this approach.

SiobhyB committed:
Set tintColor for switches in modals

SiobhyB committed:
Set tintColor for actionSheet in preseented modal

This commit also refines the code for readability.

SiobhyB commented:
@dcalhoun, I've gone ahead to make some updates that I believe catches all the UI elements in the editor. 🤞 

As a number of UI elements needed to be updated, I created a new `.editorPrimary` variable, following the same patterns found for other colours in [the `UIColor+MurielColors.swift` file](https://github.com/wordpress-mobile/WordPress-iOS/blob/d6b813b903c34d4d66ac07c5e452b378fd952e63/WordPress/Classes/Extensions/Colors%20and%20Styles/UIColor+MurielColors.swift), such as `.primary` and `.success`. Outside of that, the main changes for setting the `.tintColor` in the editor can be found in `GutenbergViewController.swift` and `PostSettingsViewController.m`. 

The changes I've made definitely come from following patterns in the codebase and Googling, rather than any significant experience in iOS development, so wide open for ideas around changes or different approaches that could be taken. Thanks in advance for looking! 🙇‍♀️

dcalhoun commented on WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift:
```diff
@@ -408,6 +408,18 @@ class GutenbergViewController: UIViewController, PostEditor, FeaturedImageDelega
+        let presentedViewActionSheet = UIView.appearance(whenContainedInInstancesOf: [UIAlertController.self])
+
+        presentedView?.tintColor = UIColor.editorPrimary
+        presentedViewSwitch.onTintColor = UIColor.editorPrimary
+        presentedViewActionSheet.tintColor = UIColor.editorPrimary
```
I believe prefixing the switch and action sheet assignments with `presentedView` is misleading, as I do not believe they are scoped to the presented view. Similar to previous issues referenced in https://github.com/wordpress-mobile/WordPress-iOS/pull/19113#issuecomment-1195965753, this is a global mutation of `UISwitch` and `UIView`/`UIAlertController`, which appears to cause "stale" color issues. 

<details><summary>Stale UISwitch color after opening the editor modal</summary>

https://user-images.githubusercontent.com/438664/181802665-79497f30-2afb-4a6c-bc54-9bcfb7f28069.MP4 

</details>

<details><summary>Stale UIAlertController color after opening the editor modal</summary>

https://user-images.githubusercontent.com/438664/181802688-c68fe3d7-4997-4ae4-998b-dcf34e957a93.MP4

</details>

Ideally we do not have to mutate the global, however, I do not know of an approach for that currently. It is possible a more experienced iOS developer could provide guidance here. 

Is the following an alternative approach for modifying `UISwitch` colors? Could a similar approach be applied here to avoid mutating the global? 

https://github.com/wordpress-mobile/WordPress-iOS/blob/7728b5a8376d7578d265f568f78fb177a03728f1/WordPress/Classes/ViewRelated/Post/PostSettingsViewController.m#L754

A workaround would be reverting the mutation whenever the modal is dismissed. I wonder if we could leverage [`dismiss`](https://developer.apple.com/documentation/uikit/uiviewcontroller/1621505-dismiss) to revert these mutations.

SiobhyB replied:
@dcalhoun, thank you so much for spotting this! 🙇‍♀️  I was able to figure out a way to [only target switches in the editor](https://github.com/wordpress-mobile/WordPress-iOS/blob/e034bc6b8833a069f49e09bfd63718aeb3ee6965/WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift#L576), by utilising `RCTModalHostViewController`, but wasn't able to follow a similar approach for action sheets.

For action sheets, I came to what feels like a really hacky workaround. Their `tintColor` is now set via the ViewController's [viewWillAppear() function](https://github.com/wordpress-mobile/WordPress-iOS/blob/43b9994672179dc1c0203f46cfce2a36e333bbb5/WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift#L373-L379) and then 'reset' via [the viewWillDisappear function](https://github.com/wordpress-mobile/WordPress-iOS/blob/43b9994672179dc1c0203f46cfce2a36e333bbb5/WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift#L388-L392). I suspect my lack of iOS expertise may be blocking me from seeing a cleaner approach.

dcalhoun commented on WordPress/Classes/ViewRelated/Post/PostSettingsViewController.m:
```diff
@@ -167,6 +168,8 @@ - (void)viewWillAppear:(BOOL)animated
     [self.navigationController setNavigationBarHidden:NO animated:NO];
     [self.navigationController setToolbarHidden:YES];
     
+    self.navigationController.view.tintColor = [UIColor murielEditorPrimary];
```
What does this mutation address? Child views?

dcalhoun replied:
Marking this as resolved, as I found the explanation you provided in the PR description. Thank you for that. 🙇🏻 

> Post Settings Screen
>
> The tintColor was set for multiple UI elements in [PostSettingsViewController](https://github.com/wordpress-mobile/WordPress-iOS/blob/4763d9c0ca69c2df63aa5e2b2b82e4928c8a23ba/WordPress/Classes/ViewRelated/Post/PostSettingsViewController.m), including its [main view](https://github.com/wordpress-mobile/WordPress-iOS/blob/4763d9c0ca69c2df63aa5e2b2b82e4928c8a23ba/WordPress/Classes/ViewRelated/Post/PostSettingsViewController.m#L144), the [subviews it's possible to navigate to](https://github.com/wordpress-mobile/WordPress-iOS/blob/4763d9c0ca69c2df63aa5e2b2b82e4928c8a23ba/WordPress/Classes/ViewRelated/Post/PostSettingsViewController.m#L171), and the [sticky toggle](https://github.com/wordpress-mobile/WordPress-iOS/blob/4763d9c0ca69c2df63aa5e2b2b82e4928c8a23ba/WordPress/Classes/ViewRelated/Post/PostSettingsViewController.m#L754).

dcalhoun reviewed (commented):
Thank you for exploring this further. You identified some great approaches. 

I encountered a few issues while testing these latest changes. 

<details><summary>Unexpected Editor UI colors</summary>

| Area | Image |
| - | - |
| Support Options | <img width="100" alt="Unexepected color in editor help support options" src="https://user-images.githubusercontent.com/438664/181803370-f25322c8-4840-431b-be54-a9258ba6fceb.PNG" /> |
| Post Preview | <img width="100" alt="Unexpected color in post preview" src="https://user-images.githubusercontent.com/438664/181803372-da23cf24-3604-46a3-b416-aa8c754264d5.PNG" /> |
| Post Category | <img width="100" alt="Unexpected color in post settings add category text input" src="https://user-images.githubusercontent.com/438664/181803378-97a8142a-3bbb-4efe-bd64-238fe792162c.PNG" /> |
| Parent Category | <img width="100" alt="Unexpected color in parent category text input" src="https://user-images.githubusercontent.com/438664/181803380-4ffb7212-45c1-4733-9ed1-2f35a42159cd.PNG" /> |
| Featured Image Media Search  | <img width="100" alt="Unexpected color in post settings featured image media search" src="https://user-images.githubusercontent.com/438664/181803379-d9384e92-cdde-48bb-a054-09719402cebe.PNG" /> |
 
</details>

I also captured a few comments on the approach, as well as a couple of stale colors I encountered after opening editor modals. 

If we find adjusting the colors for _Post Settings_ and _Editor Help_ to be too complex, we could possibly reduce the scope of what "within the editor" includes. Let me know what you think. 🙇🏻

SiobhyB committed:
Replace Post Settings' tintColor calls

With this commit, some of the tintColor calls for UI elements in Post Settings have been moved to 'WordPressAppDelegate.swift'. This is in a bid to help with overall code readability/quality.

SiobhyB committed:
Scope tintColor changes to editor

This commit, specifically, scopes the changes to switches and action sheets (accessible via the editor's block settings) to only the editor. This addresses a previous bug in which the changes were applied throughout the app.

SiobhyB committed:
Small tweak: Remove unecessary reference

SiobhyB committed:
Refactor references to Post Settings UI

Previously, the tintColor for certain UI elements within Post Settings were targeted directly from PostSettingsViewController.m. However, it's possible to target them directly from WordPressAppDelegate.swift, along with the screen's other UI elements.

This commit goes ahead to refactor these references to help with code readability/reusability.

SiobhyB committed:
Remove accidental whitespace

SiobhyB committed:
Set action sheet tintColor when editor appears

SiobhyB committed:
Update tintColor in Post Setting category screen

SiobhyB committed:
Update tintColor used in Post Settings media pick

SiobhyB committed:
Refine specificity of Post Settings tintColor call

Previously, the tintColor for the menu items in Post Settings was being targetted via the 'UIView' element. This had the undesired side effect of also changing the tintColor for other UI elements in the editor. With this commit, 'UIView' has been changed to 'UITableView' in order to more specifically target the UI elements that should be changed.

SiobhyB committed:
Add clarifying comments

SiobhyB committed:
Apply tintColor to Post Settings tag selection

SiobhyB committed:
Add clarifying comments to code

SiobhyB committed:
Merge branch 'trunk' into gutenberg/incorporate-consistent-tint-color

SiobhyB committed:
Revert accidental commit to previewController

This change was committed accidentally, and won't be a part of this PR's changes.

SiobhyB committed:
Tweaks to comment spacing/wording for clarity

SiobhyB commented:
@dcalhoun, thank you again for your diligent testing and help.

I've been able to set the colours for the main editor and Post Settings, but some of the approaches I settled on feel very hacky. I suspect my lack of iOS experience is a part of that. 

In addition, I wasn't able to find a way to customise the preview and help sections that didn't also impact their use in other areas of the app. 

What do you think about reducing the scope of this PR to include only the main editor and Post Settings screens? The other areas could then be addressed in separate PRs, depending on feedback. I'd also definitely be interested to hear thoughts on possible different, cleaner alternatives to the approaches I've taken.

dcalhoun commented on WordPress/Classes/System/WordPressAppDelegate.swift:
```diff
@@ -881,6 +881,37 @@ extension WordPressAppDelegate {
+        UIView.appearance(whenContainedInInstancesOf: [WPAddPostCategoryViewController.self]).tintColor = .editorPrimary
+        UIView.appearance(whenContainedInInstancesOf: [PostCategoriesViewController.self]).tintColor = .editorPrimary
+
+        /// It's necessary to target PostCategoriesViewController a second time to 'reset'
+        /// the UI element's `tintColor` for use in the app's Site Settings screen.
+        UIView.appearance(whenContainedInInstancesOf: [PostCategoriesViewController.self, WPSplitViewController.self]).tintColor = .primary
```
I do not understand what specific issue this addresses or how it addresses it. Would you please provide steps to replicate the issue and elaborate on how this addresses it? 

I presume the response to my above question would also apply to the [sibling reset](https://github.com/wordpress-mobile/WordPress-iOS/blob/e034bc6b8833a069f49e09bfd63718aeb3ee6965/WordPress/Classes/System/WordPressAppDelegate.swift#L912-L914) as well, but, if there is nuanced differences, will you please call that out? 

Thanks!

SiobhyB replied:
@dcalhoun, absolutely, `PostCategoriesViewController` and `WPMediaPickerViewController` are both used in two places in the app.

`PostCategoriesViewController` can be accessed via the _Post Settings_ screen and _Site Settings_ → _Categories_. So, when `(whenContainedInInstancesOf: [PostCategoriesViewController.self])` was used to change the `.tintColor`, it changed it in both screens where the controller is used. 

I couldn't see any other views in the view hierarchy when looking at the categories section of the _Post Settings_ screen. I, therefore, couldn't see a way to _only_ change the `tintColor` for that screen.

I did see some extra views when looking at the hierarchy for the _Site Settings_ → _Categories_ screen though.  That's what led me to this approach, where I change the color for both screens but then use `(whenContainedInInstancesOf: [PostCategoriesViewController.self, WPSplitViewController.self])` to 'reset' the change to _Site Settings_ → _Categories_.

Does that make sense? I agree that this doesn't seem optimal, but think my lack of iOS knowledge is again leading me down hacky paths here.

@twstokes, I wondered if you'd perhaps have ideas for cleaning this up a bit? 🙇‍♀️

dcalhoun replied:
Aha! Yes, that explanation helps immensely. I now understand the problem and solution. Thank you. 🙇🏻 

Relying upon the view hierarchy appears to be an established pattern for this file. However, the reset additions do feel a bit more brittle, as they appear to rely upon fairly generic hierarchy (e.g. `WPSplitViewController`) rather than more unique hierarchy (e.g. `PostSettingsViewController`) and require a set + reset pattern. 

Unfortunately, I do not have an alternative to propose currently. 🤔 Hopefully others will be able to provide helpful perspective for us.

twstokes replied:
👋 @SiobhyB and @dcalhoun! Thanks for the great discussion and detailed info, they've been really helpful with getting up to speed.

When I inspect the view hierarchy in Xcode with the editor open it seems to me like `AztecNavigationController` is always there. Do you think we could use that? It would have the added benefit of enveloping other views outside of `GutenbergViewController` like `PostSettingsViewController`.

Perhaps the calls could be simplified to:

```
UIView.appearance(whenContainedInInstancesOf: [AztecNavigationController.self]).tintColor = .editorPrimary
UISwitch.appearance(whenContainedInInstancesOf: [AztecNavigationController.self]).onTintColor = .editorPrimary
```

_I only did a quick smoke test so there's a really high chance I left something out or have missed something major._

![AztecNavigationController Hierarchy](https://user-images.githubusercontent.com/2092798/182748284-a5f9d7f8-6bbd-49d1-a718-0bc66e0feaac.jpg)

SiobhyB replied:
@twstokes, thank you again for helping with this! 🙇‍♀️  Your prompt made me realise there was a redundant call to `UITableView` and I went ahead to refine the `UISwitch` call, as per your suggestion, in https://github.com/wordpress-mobile/WordPress-iOS/pull/19113/commits/88689745d008e54e9a2f00b50bcc318b4156e9d2.

Refactoring the call to `UITableView` to `UIView` led to some unexpected changes in the main editor view. You'll see that the colours in the formatting toolbar, amongst a few other smaller UI elements, were changed when swapping `UITableView` with `UIView`:

| `UITableView.appearance`  | `UIView.appearance` |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/182810302-fbc62297-8d51-4233-98fd-3a3612fd82f4.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/182809729-6edfb852-d030-4154-964e-77043aec4b3e.png" width="100%"> |

In addition, I believe targeting `UIView` instead of `UITableView` here would only prevent the need to target the Post Settings' tag selection screen separately. 

The date picker, featured media selection, and category sections (specifically, the + and parent category subcreens) would still need separate calls. None of these appear to be children of `AztecNavigationController`.

Let me know if there's something I'm missing here.

The more I think about this, the more I wonder whether we're attempting to do something a bit too unusual with the way we're styling different views. I feel pretty iffy about some of the hacks in this PR. If there isn't a cleaner way to approaching some of these challenges, perhaps it'd be preferable to leave the `tintColor` in Post Settings as green and only change the editor's colours? (Though there'd still be some hacks left over to style the editor's colours.)

Looking forward to hearing if you have other ideas for tackling this, too.

twstokes replied:
> You'll see that the colours in the formatting toolbar, amongst a few other smaller UI elements, were changed when swapping UITableView with UIView

Ah, I see! I didn't spot those.

> I believe targeting UIView instead of UITableView here would only prevent the need to target the Post Settings' tag selection screen separately.

That makes sense.

> the more I wonder whether we're attempting to do something a bit too unusual with the way we're styling different views

I think we should be able to avoid the "reset" portions, at least. It's also not super common (in my experience) to set the `tintColor` via the appearance proxy, though there's nothing wrong with that. Typically I'd see it set at the view controller or view level, _but_ I also realize we're dealing with React Native so things can be different.

I'll look at this some more and report back.

twstokes replied:
@SiobhyB - would this patch be equivalent to the code currently in this section? (i.e. views that are children of `AztecNavigationController`)

**Summary:**
- Removes the resets by increasing specificity
- When there's only one `.tintColor` for an entire `UIView`, it moves the call to the corresponding view controller (it should propagate down to subviews the same way)
    - _Note:_ I didn't spend time finding the optimal spot to set `view.tintColor`, so if you go with this solution feel free to assess that. Typically somewhere in the `viewDidLoad` process works.

```
diff --git a/WordPress/Classes/System/WordPressAppDelegate.swift b/WordPress/Classes/System/WordPressAppDelegate.swift
index 91c7ec65ca..55ac1d4ab2 100644
--- a/WordPress/Classes/System/WordPressAppDelegate.swift
+++ b/WordPress/Classes/System/WordPressAppDelegate.swift
@@ -887,30 +887,8 @@ extension WordPressAppDelegate {
         /// Post Settings: Main and subscreens
         UITableView.appearance(whenContainedInInstancesOf: [AztecNavigationController.self]).tintColor = .editorPrimary
         UISwitch.appearance(whenContainedInInstancesOf: [AztecNavigationController.self]).onTintColor = .editorPrimary
-
-        /// Post Settings: Date picker
-        if #available(iOS 14.0, *) {
-            let editorSettingsDatePicker = UIView.appearance(whenContainedInInstancesOf: [SchedulingDatePickerViewController.self])
-            editorSettingsDatePicker.tintColor = .editorPrimary
-        }
-
-        /// Post Settings: Tag selection
-        UIView.appearance(whenContainedInInstancesOf: [PostTagPickerViewController.self]).tintColor = .editorPrimary
-
-        /// Post Settings: Categories selection
-        UIView.appearance(whenContainedInInstancesOf: [WPAddPostCategoryViewController.self]).tintColor = .editorPrimary
-        UIView.appearance(whenContainedInInstancesOf: [PostCategoriesViewController.self]).tintColor = .editorPrimary
-
-        /// It's necessary to target PostCategoriesViewController a second time to 'reset'
-        /// the UI element's `tintColor` for use in the app's Site Settings screen.
-        UIView.appearance(whenContainedInInstancesOf: [PostCategoriesViewController.self, WPSplitViewController.self]).tintColor = .primary
-
-        /// Post Settings: Featured Image Search
-        UIView.appearance(whenContainedInInstancesOf: [WPMediaPickerViewController.self]).tintColor = .editorPrimary
-
-        /// It's necessary to target WPMediaPickerViewController a second time to 'reset'
-        /// the UI element's `tintColor` for use app's main Media screen.
-        UIView.appearance(whenContainedInInstancesOf: [WPMediaPickerViewController.self, WPSplitViewController.self]).tintColor = .primary
+        UIView.appearance(whenContainedInInstancesOf: [AztecNavigationController.self, PostCategoriesViewController.self]).tintColor = .editorPrimary
+        UIView.appearance(whenContainedInInstancesOf: [AztecNavigationController.self, WPMediaPickerViewController.self]).tintColor = .editorPrimary
     }
 }
 
diff --git a/WordPress/Classes/ViewRelated/Post/Categories/WPAddPostCategoryViewController.m b/WordPress/Classes/ViewRelated/Post/Categories/WPAddPostCategoryViewController.m
index c7a99fb4b8..b8a15c1d1f 100644
--- a/WordPress/Classes/ViewRelated/Post/Categories/WPAddPostCategoryViewController.m
+++ b/WordPress/Classes/ViewRelated/Post/Categories/WPAddPostCategoryViewController.m
@@ -40,6 +40,7 @@ - (void)viewDidLoad
     self.tableView.sectionFooterHeight = 0.0f;
     // Set this to 0 to avoid automattic sizing of cells and getting label cell to short.
     self.tableView.estimatedRowHeight = 0.0f;
+    self.view.tintColor = [UIColor murielEditorPrimary];
 
     self.saveButtonItem = [[UIBarButtonItem alloc] initWithTitle:NSLocalizedString(@"Save", @"Save button label (saving content, ex: Post, Page, Comment, Category).")
                                                            style:[WPStyleGuide barButtonStyleForDone]
diff --git a/WordPress/Classes/ViewRelated/Post/PostTagPickerViewController.swift b/WordPress/Classes/ViewRelated/Post/PostTagPickerViewController.swift
index daecdc2af0..b86bb9f771 100644
--- a/WordPress/Classes/ViewRelated/Post/PostTagPickerViewController.swift
+++ b/WordPress/Classes/ViewRelated/Post/PostTagPickerViewController.swift
@@ -41,6 +41,7 @@ class PostTagPickerViewController: UIViewController {
         WPStyleGuide.configureTableViewColors(tableView: tableView)
 
         view.backgroundColor = .listBackground
+        view.tintColor = .editorPrimary
 
         textView.delegate = self
         // Do any additional setup after loading the view, typically from a nib.
diff --git a/WordPress/Classes/ViewRelated/Post/Scheduling/SchedulingDatePickerViewController.swift b/WordPress/Classes/ViewRelated/Post/Scheduling/SchedulingDatePickerViewController.swift
index 6d75a153ed..ce9a68378a 100644
--- a/WordPress/Classes/ViewRelated/Post/Scheduling/SchedulingDatePickerViewController.swift
+++ b/WordPress/Classes/ViewRelated/Post/Scheduling/SchedulingDatePickerViewController.swift
@@ -73,6 +73,7 @@ class SchedulingDatePickerViewController: UIViewController, DatePickerSheet, Dat
         navigationItem.setRightBarButton(doneButton, animated: false)
 
         setup(topView: chosenValueRow, pickerView: datePickerView)
+        view.tintColor = .editorPrimary
 
         setupForAccessibility()
     }
```

SiobhyB replied:
@twstokes, unfortunately, I'm seeing the green tint colours in those views that were being reset after applying the patch:

| Parent Categories  | Media Selection |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/182963243-20564a10-a4d9-4282-a43d-dab19e202167.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/182963162-52dc0cb5-3b91-4b38-b876-07a8c88ade1b.png" width="100%"> |

I don't see `AztecNavigationController.self` in the view hierarchy for either of those views, either. 🤔

twstokes replied:
> I don't see AztecNavigationController.self in the view hierarchy for either of those views, either. 🤔

Good call, I didn't drill down into the Parent Category portion to see that it's not included under  `AztecNavigationController.self`.

I didn't see the same for Media Selection, though. Perhaps we're getting there a different way? I'm adding an image block and selecting "WordPress Media Library". I'm not quite sure how that's working!

<img src="https://user-images.githubusercontent.com/2092798/182967661-606100c7-317d-48b7-9f37-577ac0c032e0.png" width="350px" />

SiobhyB replied:
> I didn't see the same for Media Selection, though. Perhaps we're getting there a different way? I'm adding an image block and selecting "WordPress Media Library". I'm not quite sure how that's working!

Aha, I see! I can confirm that I'm seeing blue when adding media via the image block, without any extra calls.

The area I'm still seeing the green tint is after navigating to Post Settings, selecting Set Featured Image, and then choosing to select an image from the Media Library. 

I'm wondering if maybe it'd be better to accept the green tint on those two screens (parent category and featured media selection) if there isn't a less hacky alternative.  

(I'll also incorporate your other suggestions for improving the overall code tomorrow, thank you for those!)

SiobhyB replied:
I've gone ahead to incorporate the suggestions from your patch in https://github.com/wordpress-mobile/WordPress-iOS/pull/19113/commits/56dab966a361844f1e30cdea8ce4d71218d0ae1b. 🙇‍♀️

dcalhoun commented on WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift:
```diff
@@ -554,12 +565,15 @@ class GutenbergViewController: UIViewController, PostEditor, FeaturedImageDelega
         view.pinSubviewToAllEdges(gutenberg.rootView)
         gutenberg.rootView.pinSubviewToAllEdges(ghostView)
+
+        UISwitch.appearance(whenContainedInInstancesOf: [RCTModalHostViewController.self]).onTintColor = .editorPrimary
```
```suggestion
        // Update the tint color of switches within React Native modals, as they require direct mutation
        UISwitch.appearance(whenContainedInInstancesOf: [RCTModalHostViewController.self]).onTintColor = .editorPrimary
```

It may be worth leaving an inline comment explaining what this accomplishes for posterity.

dcalhoun commented on WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift:
```diff
@@ -408,6 +412,13 @@ class GutenbergViewController: UIViewController, PostEditor, FeaturedImageDelega
+    override func present(_ viewControllerToPresent: UIViewController, animated flag: Bool, completion: (() -> Void)? = nil) {
+        super.present(viewControllerToPresent, animated: flag, completion: completion)
+
+        let presentedView = presentedViewController?.view
+        presentedView?.tintColor = .editorPrimary
```
```suggestion
        // Update the tint color for React Native modals when presented
        let presentedView = presentedViewController?.view
        presentedView?.tintColor = .editorPrimary
```

It may be worth leaving an inline comment explaining what this accomplishes for posterity.

SiobhyB replied:
Committed, thank you!

dcalhoun commented on WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift:
```diff
@@ -385,6 +387,8 @@ class GutenbergViewController: UIViewController, PostEditor, FeaturedImageDelega
 
     override func viewWillDisappear(_ animated: Bool) {
         super.viewWillDisappear(animated)
+        // Resets the action sheet's tintColor to `.primary` to prevent `.editorPrimary` being set as the tintColor for UI elements outside of the editor.
+        UIView.appearance(whenContainedInInstancesOf: [UIAlertController.self]).tintColor = .primary
```
The need to mutate and reset this change is unfortunate and obscure. I am curious if more experience iOS developers have idea for an alternative approach.

twstokes replied:
I wonder if we can avoid this by targeting `AztecNavigationController`. [[more context](https://github.com/wordpress-mobile/WordPress-iOS/pull/19113/files#r937295053)]

SiobhyB replied:
@twstokes, thank you for taking a look at this! 

I tried `UIView.appearance(whenContainedInInstancesOf: [UIAlertController.self, AztecNavigationController.self]).tintColor = .primary`, but it doesn't work locally for me. 

I checked the hierarchy and I believe that `AztecNavigationController` is a 'sibling' view to `UIAlertController`, rather than a parent view:

<img width="815" alt="Screenshot 2022-08-04 at 09 48 39" src="https://user-images.githubusercontent.com/2998162/182805655-6fa6be6e-b9bc-4b67-b565-763702f962b5.png">

Is there perhaps another way to target the `UIAlertController` that I'm missing?

twstokes replied:
> I checked the hierarchy and I believe that AztecNavigationController is a 'sibling' view to UIAlertController, rather than a parent view

Good catch! When I did my smoke test this code was still active so I thought the appearance proxy umbrella covered it. 👍 

> Is there perhaps another way to target the UIAlertController that I'm missing?

I'll look into this.

twstokes replied:
👋 @SiobhyB - This doesn't solve all the UIAlertController cases, but what do you think about adding `alert.view.tintColor = .editorPrimary` here to explicitly set the "More Sheet" tint color?

https://github.com/wordpress-mobile/WordPress-iOS/blob/9f4c72a6d8d374b643b20556f0101dfb50e6a035/WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController%2BMoreActions.swift#L16-L18

twstokes replied:
For `UIAlertController`s created on the RN side (e.g. Image block -> Add Image -> Choose image), is there a reason why we're not trying to enforce the tint from there (if that's possible)? Or perhaps RN has no concept of the branding colors?

twstokes replied:
If this can't be controlled / set from the RN side, I currently can't think of a better solution than what you have @SiobhyB. It's fascinating to me that the bottom sheet being presented by the editor isn't a descendant.

SiobhyB replied:
> For UIAlertControllers created on the RN side (e.g. Image block -> Add Image -> Choose image), is there a reason why we're not trying to enforce the tint from there (if that's possible)? Or perhaps RN has no concept of the branding colors?

Woah, yes, I just tested this quickly and it seems possible! It seems such an obvious solution, not sure why I went down such a complicated road with it now. 🤦‍♀️ Thanks so much @twstokes, I'll work on a Gutenberg PR.

SiobhyB replied:
I've gone ahead to create that PR at https://github.com/WordPress/gutenberg/pull/42996 and have removed the (now redundant) calls within iOS in https://github.com/wordpress-mobile/WordPress-iOS/pull/19113/commits/a67b94b00015a2b56b4301550e4f73510837f19a.

dcalhoun reviewed (commented):
Great work! I was unable to identify any unexpected outcomes while testing the latest changes. 🎉  👏🏻 

> What do you think about reducing the scope of this PR to include only the main editor and Post Settings screens?

This makes sense to me. I believe the changes proposed here provide an improved brand experience in the editor. There is less inconsistency in color. The remaining color inconsistency is buried within adjacent configuration screens. 

I left a few comments to spark discussion. I think it may be appropriate to request additional review from a more experienced iOS developer as well. That person may have ideas on alternative approaches.

SiobhyB committed:
Add clarifying comment

Co-authored-by: David Calhoun <438664+dcalhoun@users.noreply.github.com>

SiobhyB commented:
Thanks so much for your review so far @dcalhoun! 

> I think it may be appropriate to request additional review from a more experienced iOS developer as well. That person may have ideas on alternative approaches.

+1, some of the approaches I've taken feel very hacky and I suspect I arrived at them due to lack of iOS knowledge rather than a lack of cleaner alternatives. @twstokes, 👋 , would you perhaps have availability to review this PR? Thanks in advance!

SiobhyB committed:
Add clarifying comment

Co-authored-by: David Calhoun <438664+dcalhoun@users.noreply.github.com>

SiobhyB committed:
Refine/improve calls to Post Settings screen

With this commit, a redundant call to the PostSettingsViewController's 'UITableView' is removed. This call is 'covered' already by targeting AztecNavigationController's 'UITableView'.

In addition, instead of targetting PostSettingsViewController's UISwitch, this commit updates the target to AztecNavigationController.

AztecNavigationController is higher up in the view hierarchy, so will capture more UI elements.

SiobhyB committed:
Merge branch 'trunk' into gutenberg/incorporate-consistent-tint-color

twstokes commented:
One thing I noticed @SiobhyB was that Jetpack green was still coming through when selecting multiple items from the WordPress Media Library (Image Gallery block -> Add Media -> WordPress Media Library). This is partially affected by the [call here to set the tint color of the cells](https://github.com/wordpress-mobile/WordPress-iOS/blob/9f4c72a6d8d374b643b20556f0101dfb50e6a035/WordPress/Classes/System/WordPressAppDelegate.swift#L879).

The footer also had green views, so they may be explicitly set as well.

<img src="https://user-images.githubusercontent.com/2092798/182953999-da85c13f-942a-4953-b296-e234fe0df46c.png" width="350px" />

SiobhyB committed:
Remove calls to change editor action sheets

These calls are going to be replaced with direct calls in Gutenberg: https://github.com/WordPress/gutenberg/pull/42996

SiobhyB committed:
[Code quality] Refine calls to UI elements

SiobhyB commented:
@twstokes, @dcalhoun, thank you both for all your help with this. 🙇‍♀️ 

I've gone ahead to update the PR and, as it stands, it achieves the following for the Jetpack app:

* Updates the colour of UI elements within the main editor (e.g. cursor, highlight) from green to blue.
* Updates the accent of sub screens within Post Settings from green to blue.

The reset hack for the parent category screen has been left here, in order to avoid a single random screen with green remaining under Post Settings:

https://github.com/wordpress-mobile/WordPress-iOS/blob/56dab966a361844f1e30cdea8ce4d71218d0ae1b/WordPress/Classes/System/WordPressAppDelegate.swift#L889-L893

Let me know if there's a better approach for handling that.

I removed the reset option for the featured media selection, as that modal feels a bit more separate to the Post Settings screen to me and not as jarring. There's also still pending work to be done on the media selection within blocks and on the Android side, so I think it seems logical to group work together in future PRs for that.. 
 
I've summed up where we're at in [a comment on the original issue here](https://github.com/wordpress-mobile/gutenberg-mobile/issues/5047#issuecomment-1206216750) and the remaining inconsistencies are as follows:

* The media picker (available from media blocks and the Post Settings screen) still has some green.
* There is still green visible in other screens 'connected' to the editor, such as Preview, History, Help & Support, etc.

Do you think it'd make sense to ship this PR's improvements and handle the remaining issues separately? Looking forward to hearing your thoughts.

twstokes commented:
> @twstokes, @dcalhoun, thank you both for all your help with this. 🙇‍♀️
> 
> I've gone ahead to update the PR and, as it stands, it achieves the following for the Jetpack app:
> 
> * Updates the colour of UI elements within the main editor (e.g. cursor, highlight) from green to blue.
> * Updates the accent of sub screens within Post Settings from green to blue.
> 
> The reset hack for the parent category screen has been left here, in order to avoid a single random screen with green remaining under Post Settings:
> 
> https://github.com/wordpress-mobile/WordPress-iOS/blob/56dab966a361844f1e30cdea8ce4d71218d0ae1b/WordPress/Classes/System/WordPressAppDelegate.swift#L889-L893
> 
> Let me know if there's a better approach for handling that.
> 

That makes sense to me @SiobhyB!

> Do you think it'd make sense to ship this PR's improvements and handle the remaining issues separately? Looking forward to hearing your thoughts.

I think that's fair. Thanks for all your hard work and attention to detail!
